mongomock/collection.py

```diff
@@ -1754,15 +1754,40 @@ def _extend_collection(out_collection, field, expression):
                                 unwound_collection[-1], v[1:], field_item)
                     out_collection = unwound_collection
                 elif k == '$project':
-                    filter_list = ['_id']
+                    filter_list = []
+                    method = None
+                    include_id = v.get('_id')
+
                     for field, value in iteritems(v):
-                        if field == '_id' and not value:
-                            filter_list.remove('_id')
-                        elif value:
-                            filter_list.append(field)
+                        if field == '_id':
+                            continue
+                        if method is None:
+                            method = 'include' if value else 'exclude'
+                            if method == 'exclude' and include_id:
+                                raise ValueError(
+                                    "Bad projection specification, cannot include fields "
+                                    "or add computed fields during an exclusion projection: %s" % v)
+                        elif method == 'include' and not value and field != '_id':
+                            raise ValueError(
+                                "Bad projection specification, cannot exclude fields "
+                                "other than '_id' in an inclusion projection: %s" % v)
+                        elif method == 'exclude' and value:
+                            raise ValueError(
+                                "Bad projection specification, cannot include fields "
+                                "or add computed fields during an exclusion projection: %s" % v)
+                        filter_list.append(field)
+                        if value:
                             out_collection = _extend_collection(out_collection, field, value)
-                    out_collection = [{k: v for (k, v) in x.items() if k in filter_list}
-                                      for x in out_collection]
+                    if (method == 'include' and include_id is not False) or \
```

Comment:
Here you seem to assume that `method` is either `include` or `exclude`, however if there's only `_id` in the dict, then it's not defined. Can you fix this please?

Reply:
You are absolutely right. This should now be fixed by line 1760.